drivers/place/router/signal_graph.cr

```diff
@@ -0,0 +1,150 @@
+  # Given a *mod* and sets of known *inputs* and *outputs* in use on it, wire up
+  # any active edges between these based on the interfaces available.
+  protected def link(mod : Mod, inputs : Enumerable(Input), outputs : Enumerable(Output))
+    if mod.switchable? && !outputs.empty?
+      inputs.each do |input|
+        outputs.each do |output|
+          func = Edge::Func::Switch.new input.input, output.output
+          g[output.id, input.id] = Edge::Active.new mod, func
+        end
+      end
+    elsif mod.selectable?
+      inputs.each do |input|
+        output = Node::Device.new mod
+        func = Edge::Func::Select.new input.input
+        g[output.id, input.id] = Edge::Active.new mod, func
+      end
+    end
+  end
+
+  # Construct a graph from a pre-parsed configuration.
+  #
+  # *nodes* must contain the set of all signal nodes that form the device inputs
+  # and ouputs across the system. This includes those at the "edge" of the
+  # signal network (e.g. a input to a switcher) as well as inputs in use on
+  # intermediate device (e.g. a input on a display, which in turn is attached to
+  # the switcher above).
+  #
+  # *connections* declares the physical links that exist between these.
+  #
+  # Modules associated with any of these nodes are then introspected for
+  # switching, input selection and mute control based on the interfaces they
+  # expose.
+  def self.build(nodes : Enumerable(Node::Ref), connections : Enumerable({Node::Ref, Node::Ref}))
+    mod_io = Hash(Mod, {Set(Input), Set(Output)}).new do |h, k|
+      h[k] = {Set(Input).new, Set(Output).new}
+    end
+
+    siggraph = new initial_capacity: nodes.size
+
+    siggraph.insert Mute
+
+    # Create verticies for each signal node
+    nodes.each do |node|
+      siggraph.insert node
+
+      # Track device IO in use for building active edges
+      inputs, outputs = mod_io[node.mod]
+      case node
+      when Input
+        inputs << node
+      when Output
+        outputs << node
+      end
+    end
+
+    # Insert the static edges
+    connections.each { |source, dest| siggraph.connect source, dest }
+
+    # Wire up the active edges
+    mod_io.each { |mod, (inputs, outputs)| siggraph.link mod, inputs, outputs }
+
+    siggraph
+  end
+
+  # Retrieves the labelled state for *node*.
+  def [](node : Node::Ref)
+    g[node.id]
+  end
+
+  # Retrieves the labelled state for the signal node at *node_id*.
+  def [](node_id)
+    g[node_id]
+  end
+
+  # Find the signal path that connects *source* to *dest*, or `nil` if this is
+  # not possible.
+  #
+  # Provides an `Iterator` that provides labels across each node, the edge, and
+  # subsequent node.
+  def route(source : Node::Ref, destination : Node::Ref)
+    path = g.path destination.id, source.id, invert: true
+
+    return nil unless path
+
+    path.each_cons(2, true).map do |(succ, pred)|
+      {
+        g[succ],       # source
+        g[pred, succ], # edge
+        g[pred],       # next node
+      }
```

Comment:
I think a more descriptive type than a tuple would aid in use here as the positional data is pretty meaningless.
A struct or namedtuple perhaps?

Reply:
Reason for this is to enable:
https://github.com/PlaceOS/drivers/blob/7ff6e942e79ac0853abb67fc53d7b0a35a24685b/drivers/place/router/signal_graph_spec.cr#L118
So that you can keep usage of this very clean.